File: rust/src/core/src/connection.rs
use std::cell::RefCell;
use std::rc::Rc;

pub trait Connection {
    fn set_send_package_impl(&mut self, handle: Box<dyn Fn(Vec<u8>)>);
    fn send_package(&self, package: Vec<u8>);
    fn set_recv_package_impl(&mut self, handle: Box<dyn Fn(Vec<u8>)>);
    fn on_recv_package(&self, package: Vec<u8>);
}

#[derive(Default)]
pub struct DefaultConnection {
    send_package_impl: Option<Box<dyn Fn(Vec<u8>)>>,
    recv_package_impl: Option<Box<dyn Fn(Vec<u8>)>>,
}

impl DefaultConnection {
    pub fn new() -> Rc<RefCell<Self>> {
        Rc::new(RefCell::from(DefaultConnection::default()))
    }
}

impl Connection for DefaultConnection {
    fn set_send_package_impl(&mut self, handle: Box<dyn Fn(Vec<u8>)>) {
        self.send_package_impl = Some(handle);
    }

    fn send_package(&self, package: Vec<u8>) {
        if let Some(handle) = self.send_package_impl.as_ref() {
            handle(package);
        }
    }

    fn set_recv_package_impl(&mut self, handle: Box<dyn Fn(Vec<u8>)>) {
        self.recv_package_impl = Some(handle);
    }

    fn on_recv_package(&self, package: Vec<u8>) {
        if let Some(handle) = self.recv_package_impl.as_ref() {
            handle(package);
        }
    }
}

pub struct LoopbackConnection(DefaultConnection);

impl LoopbackConnection {
    pub fn new() -> Rc<RefCell<Self>> {
        let c = Rc::new(RefCell::new(LoopbackConnection(DefaultConnection::default())));
        let c_clone = c.clone();
        c.borrow_mut().0.send_package_impl = Some(
            Box::new(move |package: Vec<u8>| {
                c_clone.borrow().0.on_recv_package(package);
            })
        );
        c
    }
}

impl Connection for LoopbackConnection {
    fn set_send_package_impl(&mut self, handle: Box<dyn Fn(Vec<u8>)>) {
        self.0.set_send_package_impl(handle);
    }

    fn send_package(&self, package: Vec<u8>) {
        self.0.send_package(package);
    }

    fn set_recv_package_impl(&mut self, handle: Box<dyn Fn(Vec<u8>)>) {
        self.0.set_recv_package_impl(handle);
    }

    fn on_recv_package(&self, package: Vec<u8>) {
        self.0.on_recv_package(package);
    }
}

File: python/rpc_core/connection.py
from __future__ import annotations

import inspect
from typing import Awaitable, Callable, Protocol

SendPackageImpl = Callable[[bytes], None | Awaitable[None]]
RecvPackageImpl = Callable[[bytes], None]


class Connection(Protocol):
    def set_send_package_impl(self, handle: SendPackageImpl) -> None:
        ...

    def send_package(self, package: bytes | bytearray | memoryview) -> None | Awaitable[None]:
        ...

    def set_recv_package_impl(self, handle: RecvPackageImpl) -> None:
        ...

    def on_recv_package(self, package: bytes | bytearray | memoryview) -> None:
        ...


class DefaultConnection:
    def __init__(self) -> None:
        self._send_package_impl: SendPackageImpl | None = None
        self._recv_package_impl: RecvPackageImpl | None = None

    def set_send_package_impl(self, handle: SendPackageImpl) -> None:
        self._send_package_impl = handle

    def send_package(self, package: bytes | bytearray | memoryview) -> None | Awaitable[None]:
        if self._send_package_impl is None:
            return None
        result = self._send_package_impl(bytes(package))
        return result if inspect.isawaitable(result) else None

    def set_recv_package_impl(self, handle: RecvPackageImpl) -> None:
        self._recv_package_impl = handle

    def on_recv_package(self, package: bytes | bytearray | memoryview) -> None:
        if self._recv_package_impl is not None:
            self._recv_package_impl(bytes(package))


class LoopbackConnection:
    @staticmethod
    def create() -> tuple[DefaultConnection, DefaultConnection]:
        c1 = DefaultConnection()
        c2 = DefaultConnection()
        c1.set_send_package_impl(lambda package: c2.on_recv_package(package))
        c2.set_send_package_impl(lambda package: c1.on_recv_package(package))
        return c1, c2
